WPMedia/fusion-news-theme-blocks
# ASUB-8002 Update payment method

**If you have not filled out the checklist below, the pr is not ready for review.**

## Description

We are adding a new Subs blocks, this block will allow us to update a payment method. Currently, our checkout block allows payments with Stripe Intents and Paypal (this last one in progress).
The Payment update is supported from 
- stripe Intents to Stripe Intents
- Paypal to Stripe Intents
**Stripe Intents to PayPal is not supported in our BE**
This is useful when the renewal fails because the card expired,  or when the user wants to change the payment method to continue with their subscriptions.
Stripe documentation if you want to use different card number >> https://stripe.com/docs/testing
e.g. 4000000000003220
4000008400001280

https://github.com/WPMedia/arc-themes-blocks/assets/54566275/b289d500-c5cc-4617-807c-7950f10995fb


We added some changes to the Identity block since we saw some parameters were being missed.

## Jira Ticket

https://arcpublishing.atlassian.net/browse/ASUB-8002

## Acceptance Criteria

_copy from ticket_

## Test Steps

- Add test steps a reviewer must complete to test this PR

1. Checkout this branch `git checkout ASUB-8002_UpdatePaymentMethod`
2. Run fusion repo with linked blocks `npx fusion start -l @wpmedia/subscriptions-block,@wpmedia/identity-block`
3. Create a page in fusion named `updatepayment`, and include `Subscriptions Update payment method - Arc Block`
3. Create a custom block with the following info, create a page, and include the block. 
This block will grab the active subscriptions for a logged-in user, and for each subscription, you will have a link text from which you can fall on the `update payment method` page.  
pre-requisite: The user should have active paid subscriptions, In case you don't have active subscriptions you will need to go through the checkout process. 

```
import React, { useState, useEffect } from "react";
import PropTypes from "@arc-fusion/prop-types";

import { Heading, Link, useIdentity, Stack } from "@wpmedia/arc-themes-components";
import useSales from "../../components/useSales";

const SubscriptionList = ({ customFields }) => {
	const { loginUrl } = customFields;

	const [subscriptionList, setSubscriptionList] = useState([]);

	const { Identity } = useIdentity();
	const { Sales } = useSales();

	const getIsLoggedIn = async () => {
		const isLoggedIn = await Identity.isLoggedIn();

        if (isLoggedIn) {
			getAllSubscriptions().then((subs) => {
				setSubscriptionList(subs);
			});
		} else {
			const updatePaymentUrl = new URL(window.location.href);
			window.location.href = `${loginUrl}${loginUrl.indexOf("?") != -1 ? "&" : "?"}redirect=${
				updatePaymentUrl?.pathname
			}${updatePaymentUrl.search}`;
		}
	};

	const getAllSubscriptions = async () => {
		const listSubs = await Sales.getAllActiveSubscriptions();
		return listSubs;
	};

	useEffect(() => {
		getIsLoggedIn();
	}, [Identity]);

	return (
		<section>
			<Heading>Active Subscriptions</Heading>
			<Stack>
				{subscriptionList.map((sub) => {
					return (
						<Link
							href={`http://localhost/pf/updatepayment/?_website=the-gazette&paymentMethodID=${sub?.paymentMethod?.paymentMethodID}`}
						>{`${sub.productName} - ${sub.subscriptionID} - ${sub?.paymentMethod?.paymentPartner} - ${sub?.paymentMethod?.lastFour} `}</Link>
					);
				})}
			</Stack>
		</section>
	);
};

SubscriptionList.propTypes = {
	customFields: PropTypes.shape({
		loginUrl: PropTypes.string.tag({
			label: "Log In link URL",
			defaultValue: "/account/login/",
		}),
	}),
};

SubscriptionList.label = "Subscriptions active Subscriptions - Custom Block";

export default SubscriptionList;
```


## Effect Of Changes

### Before

_Example: When I clicked the search button, the button turned red._

[include screenshot or gif or link to video, storybook would be awesome]

### After

_Example: When I clicked the search button, the button turned green._

[include screenshot or gif or link to video, storybook would be awesome]

## Dependencies or Side Effects

_Examples of dependencies or side effects are:_

- Manifest pr for new block:
- Feature pack pr for local development:
- How to deploy to news: https://arcpublishing.atlassian.net/wiki/spaces/TI/pages/3138682964/News+Theme+2.0+Migration+Development+Deployment+Notes
- Additional settings that will be required in the blocks.json
- Changes to the custom fields which will require users to reconfigure features
- Dependency on another PR that needs to be merged first

## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._

        please explain why (so that we can fix it whenever it gets refactored).

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._
